"py -i" global namespace bug

Peter.Lee@deshaw.com reported this bug below.

This is reproducible only when ipython startup config files don't already enable pyflyby.

```
globals() isn't locals() in the session below.

pylogusr@dirac15.toa(nyc):~> py -i /dev/null

In [1]: import random

In [2]: def f(): random

In [3]: f()
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-3-c43e34e6d405> in <module>()
----> 1 f()

<ipython-input-2-0708d8019c7a> in f()
----> 1 def f(): random

NameError: global name 'random' is not defined


```

This is what it breaks:

```
__________________________________________________________________ test_py_i_interactive_1 ___________________________________________________________________

tmp = <tests.test_interactive._TmpFixture object at 0x11e3a7f28>

    @retry
    def test_py_i_interactive_1(tmp):
        # Test that 'py -i' initializes IPython before running the commandline
        # code.
        # (The converse is tested by test_py.py:test_no_ipython_for_eval_1.)
        writetext(tmp.dir / "m32622167.py", """
            import pyflyby
            ipython_app = pyflyby._interactive._get_ipython_app()
        """)
        ipython("""
            In [1]: bool(m32622167.ipython_app)
            Out[1]: True
>       """, prog="py", args=['-i', 'import m32622167'], PYTHONPATH=tmp.dir)
E       Failed: Expected:
E            b'In [1]: bool(m32622167.ipython_app)'
E            b'Out[1]: True'
E       Result:
E            b'In [1]: bool(m32622167.ipython_app)'
E            b'[PYFLYBY] import m32622167'
E            b'Out[1]: True'
E       Diff:
E            In [1]: bool(m32622167.ipython_app)
E          + [PYFLYBY] import m32622167
E            Out[1]: True

tests/test_interactive.py:3202: Failed
```

But I think we can just `.update` the namespace and get the desired effect, without making globals and locals separate dictionaries.

--- a/lib/python/pyflyby/_interactive.py
+++ b/lib/python/pyflyby/_interactive.py
@@ -293,7 +293,7 @@ def start_ipython_with_autoimporter(argv=None, app=None, _user_ns=None):
         # be the same dict.  We should create a true ModuleType anyway even if
         # not using IPython.  We might need to resort to advising
         # init_create_namespaces etc. depending on IPython version.
-        if app.shell:
+        if getattr(app, 'shell', None) is not None:
             app.shell.user_ns.update(_user_ns)
         else:
             app.user_ns = _user_ns

--- a/lib/python/pyflyby/_interactive.py
+++ b/lib/python/pyflyby/_interactive.py
@@ -293,7 +293,7 @@ def start_ipython_with_autoimporter(argv=None, app=None, _user_ns=None):
         # be the same dict.  We should create a true ModuleType anyway even if
         # not using IPython.  We might need to resort to advising
         # init_create_namespaces etc. depending on IPython version.
-        if app.shell.user_ns:
+        if app.shell:
             app.shell.user_ns.update(_user_ns)
         else:
             app.user_ns = _user_ns

--- a/lib/python/pyflyby/_interactive.py
+++ b/lib/python/pyflyby/_interactive.py
@@ -293,7 +293,7 @@ def start_ipython_with_autoimporter(argv=None, app=None, _user_ns=None):
         # be the same dict.  We should create a true ModuleType anyway even if
         # not using IPython.  We might need to resort to advising
         # init_create_namespaces etc. depending on IPython version.
-        app.user_ns = _user_ns
+        app.shell.user_ns.update(_user_ns)
     return _initialize_and_start_app_with_autoimporter(app, argv)
 
 

--- a/lib/python/pyflyby/_interactive.py
+++ b/lib/python/pyflyby/_interactive.py
@@ -293,7 +293,10 @@ def start_ipython_with_autoimporter(argv=None, app=None, _user_ns=None):
         # be the same dict.  We should create a true ModuleType anyway even if
         # not using IPython.  We might need to resort to advising
         # init_create_namespaces etc. depending on IPython version.
-        app.shell.user_ns.update(_user_ns)
+        if app.shell.user_ns:
+            app.shell.user_ns.update(_user_ns)
+        else:
+            app.user_ns = _user_ns
     return _initialize_and_start_app_with_autoimporter(app, argv)
 
 

